File: src/main/java/net/volcano/jdacommands/model/command/arguments/implementation/argument/UserArgument.java
package net.volcano.jdacommands.model.command.arguments.implementation.argument;

import lombok.experimental.SuperBuilder;
import net.dv8tion.jda.api.entities.User;
import net.volcano.jdacommands.exceptions.command.parsing.InvalidArgumentsException;
import net.volcano.jdacommands.model.command.arguments.CommandArgument;
import net.volcano.jdacommands.model.command.arguments.implementation.ArgumentParsingData;
import net.volcano.jdautils.utils.UserUtil;

@SuperBuilder
public class UserArgument extends CommandArgument<User> {
	
	protected final boolean defaultToCaller;
	
	@Override
	public User parseValue(ArgumentParsingData data) throws InvalidArgumentsException {
		User user = UserUtil.findUser(data.getArg(), data.event.getJDA(), data.event.isFromGuild() ? data.event.getGuild() : null);
		if (user != null || nullable) {
			return user;
		} else {
			throw new InvalidArgumentsException(data, String.format("User \"%s\" not found.", data.getArg()));
		}
	}
	
}

File: src/main/java/net/volcano/jdacommands/model/command/arguments/implementation/argument/UserArgument.kt
package net.volcano.jdacommands.model.command.arguments.implementation.argument

import net.dv8tion.jda.api.entities.User
import net.volcano.jdacommands.exceptions.command.parsing.InvalidArgumentsException
import net.volcano.jdacommands.model.command.arguments.CommandArgument
import net.volcano.jdacommands.model.command.arguments.implementation.ArgumentParsingData
import net.volcano.jdautils.utils.UserUtil.findUser

class UserArgument(
	/**
	 * If the argument should default to the command caller in-case another user is not found.
	 */
	private val defaultToCaller: Boolean = false
) : CommandArgument<User?>() {

	@Throws(InvalidArgumentsException::class)
	override fun parseValue(data: ArgumentParsingData): User? {
		val user = findUser(data.arg, data.event.jda, if (data.event.isFromGuild) data.event.guild else null)

		return if (user != null)
			user
		else if (defaultToCaller)
			data.event.author
		else if (nullable)
			null
		else
			throw InvalidArgumentsException(data, "User \"${data.arg}\" not found.")
	}

}